File: src/main/java/com/github/kindrat/cassandra/client/ui/ToggleSwitch.java
package com.github.kindrat.cassandra.client.ui;

import javafx.beans.property.SimpleBooleanProperty;
import javafx.geometry.Pos;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.layout.HBox;

import static javafx.application.Platform.runLater;

public class ToggleSwitch extends HBox {
    private final Label label = new Label();
    private final Button button = new Button();

    private SimpleBooleanProperty switchedOn = new SimpleBooleanProperty(false);

    public ToggleSwitch(String on, String off) {
        label.setText(off);
        getChildren().addAll(label, button);
        button.setOnAction((e) -> switchedOn.set(!switchedOn.get()));
        label.setOnMouseClicked((e) -> switchedOn.set(!switchedOn.get()));

        setStyle();
        bindProperties();

        onEnabled(() -> {
            label.setText(on);
            setStyle("-fx-background-color: lightgrey; -fx-text-fill:black; -fx-background-radius: 4;");
            label.toFront();
        });
        onDisabled(() -> {
            label.setText(off);
            setStyle("-fx-background-color: lightgrey; -fx-text-fill:black; -fx-background-radius: 4;");
            button.toFront();
        });
    }

    public SimpleBooleanProperty switchOnProperty() {
        return switchedOn;
    }

    public void onEnabled(Runnable action) {
        switchedOn.addListener((a, oldValue, newValue) -> {
            if (newValue) {
                runLater(action);
            }
        });
    }

    public void onDisabled(Runnable action) {
        switchedOn.addListener((a, oldValue, newValue) -> {
            if (!newValue) {
                runLater(action);
            }
        });
    }

    private void setStyle() {
        //Default Width
        setWidth(100);
        label.setAlignment(Pos.CENTER);
        setStyle("-fx-background-color: lightgrey; -fx-text-fill:black; -fx-background-radius: 4;");
        setAlignment(Pos.CENTER_LEFT);
    }

    private void bindProperties() {
        label.prefWidthProperty().bind(widthProperty().divide(2));
        label.prefHeightProperty().bind(heightProperty());
        button.prefWidthProperty().bind(widthProperty().divide(2));
        button.prefHeightProperty().bind(heightProperty());
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/ui/fx/component/ToggleSwitch.kt
package com.github.kindrat.cassandra.client.ui.fx.component

import javafx.application.Platform
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.value.ObservableValue
import javafx.event.EventHandler
import javafx.geometry.Pos
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.layout.HBox

class ToggleSwitch(on: String, off: String) : HBox() {

    private val label = Label()
    private val button = Button()
    private val switchedOn = SimpleBooleanProperty(false)

    fun switchOnProperty(): SimpleBooleanProperty {
        return switchedOn
    }

    fun onEnabled(action: Runnable?) {
        switchedOn.addListener { _: ObservableValue<out Boolean>, _: Boolean, newValue: Boolean ->
            if (newValue) {
                Platform.runLater(action)
            }
        }
    }

    fun onDisabled(action: Runnable?) {
        switchedOn.addListener { _: ObservableValue<out Boolean>, _: Boolean, newValue: Boolean ->
            if (!newValue) {
                Platform.runLater(action)
            }
        }
    }

    private fun setStyle() {
        //Default Width
        width = 100.0
        label.alignment = Pos.CENTER
        style = "-fx-background-color: lightgrey; -fx-text-fill:black; -fx-background-radius: 4;"
        alignment = Pos.CENTER_LEFT
    }

    private fun bindProperties() {
        label.prefWidthProperty().bind(widthProperty().divide(2))
        label.prefHeightProperty().bind(heightProperty())
        button.prefWidthProperty().bind(widthProperty().divide(2))
        button.prefHeightProperty().bind(heightProperty())
    }

    init {
        label.text = off
        children.addAll(label, button)
        button.onAction = EventHandler { switchedOn.set(!switchedOn.get()) }
        label.onMouseClicked = EventHandler { switchedOn.set(!switchedOn.get()) }
        setStyle()
        bindProperties()
        onEnabled(Runnable {
            label.text = on
            style = "-fx-background-color: lightgrey; -fx-text-fill:black; -fx-background-radius: 4;"
            label.toFront()
        })
        onDisabled(Runnable {
            label.text = off
            style = "-fx-background-color: lightgrey; -fx-text-fill:black; -fx-background-radius: 4;"
            button.toFront()
        })
    }
}